### PR TITLE
Progress of july 31

### PR DESCRIPTION
**Pull request by Saitster**
Incomplete. Modified what you said, corrected mistakes in validation, implementing calculation of three quantities related to z (in the way)

### CODE REVIEW

File: app/Controller/LucasUzawaController.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+            'alpha' => rand(1, 99) / 100,
+            't' => rand(1, 100),
+            'k0' => rand(1, 100),
+            'h0' => rand(1, 100),
+            'c0' => rand(1, 100) * 1000,
+            'u0' => rand(1, 99) / 100];
+
+            //Validate input. If there is an error, the error message is returned.
+            $error_message = ValidateInputVars::validate($exovars);
+            if(isset($error_message) and $error_message !== ''){
```
Comment (stephen1807) [inline app/Controller/LucasUzawaController.php]: You can simplify to `!empty($error_message)`.

File: app/Controller/LucasUzawaController.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
+<?php
+    //A class where functions for validating input are separated
+    App::uses('dBug', 'Lib');
+    App::uses('ValidateInputVars', 'Lib');
+
+    class LucasUzawaController extends AppController {
+        function index(){
```
Comment (stephen1807) [inline app/Controller/LucasUzawaController.php]: Don't forget access modifiers: `public`, `private`, `protected`

File: app/Controller/LucasUzawaController.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            $z0 =
+                    ((($exovars['alpha'] - 1) * $exovars['A'] * $exovars['k0'] * $exovars['rho'])
+                    / ($exovars['delta'] * $exovars['u0'] * ($exovars['k0'] * $exovars['rho'] - $exovars['c0'])))
+                ** (1 / ($exovars['alpha'] - 1));
+
+            $exovars_with_zs = array_merge($exovars, ['z_star' => $z_star], ['z0' => $z0]);
+
+            //new dBug($exovars_with_zs);
+            //Calculate z as a function of t
+            function z($param_with_zs){
```
Comment (stephen1807) [inline app/Controller/LucasUzawaController.php]: Bad function naming. Generally, the form is `verb` + `noun`

File: app/Controller/LucasUzawaController.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                return
```
Comment (stephen1807) [inline app/Controller/LucasUzawaController.php]: You must not directly return such long calculation.

File: app/Lib/ValidateInputVars.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    if(is_null($val) or $val === ''){
+                        throw new Exception('Please fill in all the input fields.');
+                    }
+
+                    //Note for the future myself: I tried ctype_digit(strval($exo_var)) but then 0.5 -> '0.5' -> '.' isn't a num -> '0.5' isn't a num
+                    if(!is_numeric($val)){
+                        throw new Exception('Please enter a number (half-width character).');
+                    }
+
+                    if($val <= 0){
```
Comment (stephen1807) [inline app/Lib/ValidateInputVars.php]: If `0` is considered wrong, then you can just simplify validation to just `empty()` and `is_numeric()`

File: app/Lib/ValidateInputVars.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
+<?php
+    /**
+     * Load useful scripts to make debugging easier
+     * https://github.com/ospinto/dBug
+     */
+    App::uses('dBug', 'Lib');
+
+    class ValidateInputVars{
+            public static function validate($v){
```
Comment (stephen1807) [inline app/Lib/ValidateInputVars.php]: Parameter type? `array`, `int`, etc.

File: app/Lib/ValidateInputVars.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            try{
+                foreach($v as $name => $val){
```
Comment (stephen1807) [inline app/Lib/ValidateInputVars.php]: Try to loop the `Validate`, not put the loop inside it. I.e. perform them one by one, and different function / class for each validation.

Then, if you separate the function / class, you can catch error independently and return their own exception messages, not writing such long exception messages like down below

File: app/Lib/ValidateInputVars.php [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                 * Thinking carefully, this is multivalued
+                 * (- 1.3) ** 12.5
+                 * 
+                 * In order to avoid such calculation the following cases must be excluded
+                 */
+                if(
+                        (1 / ($v['alpha'] - 1)) % 1!= 0
+                    and
+                        0 > 
+                            (1 - exp((1 - 1 / $v['alpha']) * $v['delta'] * $v['t']))
```
Comment (stephen1807) [inline app/Lib/ValidateInputVars.php]: Don't make such long calculation in 1 line! Separate them to multiple lines

**Comment by Saitster** [comment]
Stop funny name branch

**Pull request closed without merging**